core/src/executor/validate.rs

```diff
@@ -74,39 +79,93 @@ impl UniqueConstraint {
 }
 
 pub async fn validate_unique(
-    storage: &impl Store,
+    storage: &dyn Store,
     table_name: &str,
     column_validation: ColumnValidation,
     row_iter: impl Iterator<Item = &Row> + Clone,
 ) -> Result<()> {
-    let columns = match column_validation {
-        ColumnValidation::All(column_defs) => fetch_all_unique_columns(&column_defs),
-        ColumnValidation::SpecifiedColumns(column_defs, specified_columns) => {
-            fetch_specified_unique_columns(&column_defs, &specified_columns)
+    enum Columns {
+        /// key index
+        PrimaryKeyOnly(usize),
+        /// `[(key_index, table_name)]`
+        All(Vec<(usize, String)>),
+    }
+
+    let columns = match &column_validation {
+        ColumnValidation::All(column_defs) => {
+            let primary_key_index = column_defs
+                .iter()
+                .enumerate()
+                .find(|(_, column_def)| {
+                    column_def
+                        .options
+                        .iter()
+                        .any(|ColumnOptionDef { option, .. }| {
+                            matches!(option, ColumnOption::Unique { is_primary: true })
+                        })
+                })
+                .map(|(i, _)| i);
+            let other_unique_column_def_count = column_defs
+                .iter()
+                .filter(|column_def| {
+                    column_def
+                        .options
+                        .iter()
+                        .any(|ColumnOptionDef { option, .. }| {
+                            matches!(option, ColumnOption::Unique { is_primary: false })
+                        })
+                })
+                .count();
+
+            match (primary_key_index, other_unique_column_def_count) {
+                (Some(primary_key_index), 0) => Columns::PrimaryKeyOnly(primary_key_index),
+                _ => Columns::All(fetch_all_unique_columns(column_defs)),
+            }
         }
+        ColumnValidation::SpecifiedColumns(column_defs, specified_columns) => Columns::All(
+            fetch_specified_unique_columns(column_defs, specified_columns),
+        ),
```

Comment:
`SpecifiedColumns` is for `UPDATE` Statement, but PrimaryKey cannot be updated, so it is not checked.